### PR TITLE
Shard fs cache keys

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Shard fs cache keys similar to `OpenedFileCache`.

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit f51f6d13571c1685c826a6cab11d289aaec69a94 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56194/f51f6d13571c1685c826a6cab11d289aaec69a94/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6718630620">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56194/f51f6d13571c1685c826a6cab11d289aaec69a94/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Review by nickitat — APPROVED** [review]

File: src/Interpreters/Cache/Metadata.cpp [hunk 1]
```diff
@@ -168,12 +167,18 @@ String CacheMetadata::getPathForKey(const Key & key) const
...
-CacheMetadataGuard::Lock CacheMetadata::lockMetadata() const
+CacheMetadataGuard::Lock CacheMetadata::MetadataBucket::lock() const
 {
     ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::FilesystemCacheLockMetadataMicroseconds);
     return guard.lock();
 }
 
+CacheMetadata::MetadataBucket & CacheMetadata::getMetadataBucket(const Key & key)
+{
+    const auto bucket = key.key % buckets_num;
```
Comment (nickitat) [inline src/Interpreters/Cache/Metadata.cpp:178]: `key.key` is `sipHash(path)`, right?

Comment (kssenii) [reply src/Interpreters/Cache/Metadata.cpp:178]: yes

**Comment by kssenii** [comment]
> Integration tests (asan) [6/6] — fail: 1

test_delayed_replica_failover/

> Stateless tests (release, s3 storage) [1/2] — fail: 1

01171_mv_select_insert_isolation_long

**Pull request merged**